### PR TITLE
Fixed #393 - Make Docker available to 32 bit apps

### LINKED ISSUE #393: Visual Studio reports "Docker is not installed" after docker-ce installation on Windows Server 2019
**What is your question?**

Hi there; I support some legacy .Net Framework applications and we had met with some degree of success putting them into containers and migrating them to the cloud. For our build environment, we have been leveraging Windows Server 2019 with Visual Studio 2019 Build Tools and Docker installed through the [OneGet MicrosoftDockerProvider](https://github.com/OneGet/MicrosoftDockerProvider) which is what the Microsoft documentation had a while back.

The OneGet provider has been deprecated since May, so we tried installing through the script provided in this repository for docker-ce; however, Visual Studio build tools complains that it can't locate Docker, even though it is, in fact, installed.

```
PS C:\Users\Administrator\git\myproject\Web> docker version
Client:
 Version:           24.0.2
 API version:       1.43
 Go version:        go1.20.4
 Git commit:        cb74dfc
 Built:             Thu May 25 21:53:15 2023
 OS/Arch:           windows/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          24.0.2
  API version:      1.43 (minimum version 1.24)
  Go version:       go1.20.4
  Git commit:       659604f9
  Built:            Thu May 25 21:52:13 2023
  OS/Arch:          windows/amd64
  Experimental:     false

PS C:\Users\Administrator\git\myproject\Web> msbuild.exe .\web.csproj /t:containerbuild /p:Configuration=Release
Microsoft (R) Build Engine version 16.11.2+f32259642 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 7/3/2023 8:26:49 PM.
Project "C:\Users\Administrator\git\myproject\Web\web.csproj" on node 1 (containerbuild target(s)).
ContainerVerifyDockerInstallation:
  'docker' is not recognized as an internal or external command,
  operable program or batch file.
C:\Users\Administrator\.nuget\packages\microsoft.visualstudio.azure.containers.tools.targets\1.11.1\build\Container.tar
gets(97,5): error : Docker is not installed or is not in the current PATH. [C:\Users\Administrator\git\myproject\Web\
web.csproj]
Done Building Project "C:\Users\Administrator\git\myproject\Web\web.csproj" (containerbuild target(s)) -- FAILED.


Build FAILED.

"C:\Users\Administrator\git\myproject\Web\web.csproj" (containerbuild target) (1) ->
(ContainerVerifyDockerInstallation target) ->
  C:\Users\Administrator\.nuget\packages\microsoft.visualstudio.azure.containers.tools.targets\1.11.1\build\Container.t
argets(97,5): error : Docker is not installed or is not in the current PATH. [C:\Users\Administrator\git\myproject\We
b\web.csproj]

    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:00.47
```

Is there something missing from the install script that Visual Studio is looking for? It looks like the container target is running `docker --version` which runs fine in the same PowerShell prompt.

```
PS C:\Users\Administrator\git\myproject\Web> docker --version
Docker version 24.0.2, build cb74dfc
```

`where.exe` also reports that `docker` is in `C:\Windows\System32`:

```
PS C:\Users\Administrator\git\myproject\Web> where.exe docker
C:\Windows\System32\docker.exe
```

### PR DESCRIPTION
**Pull request by iancward**
When using WOW64, C:\Windows\System32 actually points to C:\Windows\SysWow64.

### CODE REVIEW

**Comment by iancward** [comment]
@ntrappe-msft @michbern-ms and chance you folks could review this PR?

**Comment by iankingori** [comment]
@iancward please provide some additional context on the specific scenario you are solving and an approach to test this.

**Comment by iancward** [comment]
Hi [iankingori](https://github.com/iankingori), the context is in the referenced issue: #393.

Namely, the docker-ce helper script in this repository installs docker.exe by copying it directly to `C:\Windows\System32`. When you're running a 32 bit application (e.g. `msbuild.exe` for [Visual Studio 2019 or older](https://learn.microsoft.com/en-us/visualstudio/msbuild/whats-new-msbuild-17-0?view=vs-2022#64-bit)), the `%windir%\System32` directory is [mapped](https://learn.microsoft.com/en-us/windows/win32/winprog64/file-system-redirector) to `%windir%\SysWOW64` and therefore any attempts to call `docker.exe` installed by this helper script by a 32-bit application will fail, because it doesn't exist in the path.

TL;DR: this PR makes `docker.exe` available for 32 bit applications.

**Comment by iancward** [comment]
Hi @iankingori @ntrappe-msft any updates on this?

**Comment by iancward** [comment]
Hi @iankingori @ntrappe-msft any updates on this?

**Comment by iancward** [comment]
Hi @iankingori @ntrappe-msft any updates on this?

**Comment by iancward** [comment]
Hi @iankingori @ntrappe-msft @michbern-ms any updates on this?

**Comment by michbern-ms** [comment]
Hi, @iancward - sorry for the delay.  Some of our team members were impacted by job eliminations and we're trying to get our house in order ... but I see you've been waiting a long time.  Let me see what I can do.

**Comment by iancward** [comment]
Hi @michbern-ms any updates on this?

**Comment by fjs4** [comment]
Hi @iancward, it appears you're trying to work around the fact that by default trying to run a program from \windows\system32 when the parent is a 32-bit app, will actually try to launch from \windows\syswow64.  Your proposed fix to put the 64-bit version of Docker in that location would work around the problem.  However, the intention of that directory is to only container 32-bit binaries.

What is the parent 32-bit process?  And how is it launched?  Rather than trying to run the program from \windows\system32, you could instead try to run \windows\sysnative\Docker.exe.  If the process interpreting that path is a 32-bit process, it will launch the 64-bit version of Docker.exe from \windows\system32 directory.

Will that approach work for you?

**Comment by iancward** [comment]
Hi @fjs4, as mentioned above, I'm using msbuild.exe, which in some circumstances is a 32-bit application. I'm not calling docker.exe directly, but it's being called by msbuild.exe when it's building a project with Docker targets, and it must exist in the Path, or the build will fail.

When you install Docker Desktop on Windows, it is installed to `C:\Program Files\Docker\` and the bin folder is put in the path. However, if you're on Windows Server and use the helper script in this repo, instead of putting the Docker in `C:\Program Files\Docker` or elsewhere, it downloads a zip file of binaries and places `docker.exe` directly in `C:\Windows\System32` which is inaccessible to 32-bit applications; you will receive an error that `docker.exe` is not found.

At the end of the day, `docker.exe` needs to be accessible in the Path 32-bit applications or we can't build our ASP.Net projects in CI/CD via msbuild.exe. I've successfully tested the changes in this PR and in fact have been using similar changes in our production environment

**Comment by fjs4** [comment]
Thanks for the additional clarity @iancward.  I assume you're using an Exec command in your msbuild project file?  If so, my suggestion was that you specify the path in the command, something like this:

```
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <Target Name="RunMyExecutable">
    <Exec Command="c:\windows\SysNative\Docker.exe <arguments>" />
  </Target>
</Project>
```

When invoked from a 32-bit build process, "SysNative" should force the system to launch the child process from \windows\system32.

Let me know if that helps.  If not, could you share a snippet of your project file (to help me better understand).

**Comment by iancward** [comment]
Hi @fjs4,
I didn't add the docker configuration directly, I simply added Docker support to the .Net Framework solution following the documentation provided [here](https://learn.microsoft.com/en-us/visualstudio/containers/add-container-support?view=vs-2019).

And then on my build server (Windows Server 2019), I installed docker via the helper script in this repo. And we call `msbuild.exe` with the `/t:containerbuild` argument.

Instead of having to modify all of our projects, it would be nice if the script provided in this repo would work out of the box.

Maybe instead of copying the `docker.exe` file directly to `C:\Windows\System32`, it would be appropriate to add it to a `C:\Program Files\Docker\` directory and then add it to the PATH?

**Comment by iancward** [comment]
@fjs4, @michbern-ms

Any updates on this?

**Comment by iancward** [comment]
@fjs4, @michbern-ms, @ntrappe-msft

Any updates on this?

**Comment by iancward** [comment]
@fjs4 @ntrappe-msft @michbern-ms and updates on this?

**Comment by iancward** [comment]
@fjs4 @ntrappe-msft @michbern-ms and updates on this?

**Comment by iancward** [comment]
@fjs4 @ntrappe-msft @michbern-ms and updates on this?

**Comment by iancward** [comment]
@fjs4 @ntrappe-msft @michbern-ms and updates on this?

@AvrumFeldman is this something you can review?

**Comment by smiezah-msft** [comment]
Thanks for raising this issue @iancward. After reviewing the behavior, it appears the underlying challenge is related to WoW64 filesystem redirection on 64-bit Windows. When a 32-bit process attempts to locate or launch Docker, Windows may redirect certain path lookups, causing the process to resolve a different location than expected.
 
Rather than placing the 64-bit Docker binaries in SysWOW64 (which is intended for 32-bit binaries), our recommendation would be to install Docker in a location outside of the redirected Windows system directories, such as Program Files, and ensure that location is explicitly added to the system PATH environment variable. This allows both 32-bit and 64-bit applications to locate and invoke the Docker executable consistently, without relying on filesystem redirection behavior.
 
In short, the issue is not that Docker itself is unavailable to 32-bit applications, but that the application's path resolution is affected by WoW64 redirection. Installing Docker in a non-redirected location and referencing it through the PATH provides a more robust and supportable solution.

**Comment by iancward** [comment]
> Thanks for raising this issue @iancward. After reviewing the behavior, it appears the underlying challenge is related to WoW64 filesystem redirection on 64-bit Windows. When a 32-bit process attempts to locate or launch Docker, Windows may redirect certain path lookups, causing the process to resolve a different location than expected.
> 
> Rather than placing the 64-bit Docker binaries in SysWOW64 (which is intended for 32-bit binaries), our recommendation would be to install Docker in a location outside of the redirected Windows system directories, such as Program Files, and ensure that location is explicitly added to the system PATH environment variable. This allows both 32-bit and 64-bit applications to locate and invoke the Docker executable consistently, without relying on filesystem redirection behavior.
> 
> In short, the issue is not that Docker itself is unavailable to 32-bit applications, but that the application's path resolution is affected by WoW64 redirection. Installing Docker in a non-redirected location and referencing it through the PATH provides a more robust and supportable solution.

I think that makes sense @smiezah-msft and [what I originally proposed](https://github.com/microsoft/Windows-Containers/issues/393#issuecomment-1624156839). Since I'm just a visitor in your house, do you have an opinion on where the file(s) should be extracted to? It looks like Docker Desktop installs itself to `C:\Program Files\Docker\Docker`, I could update the MR to save to `C:\Program Files\Docker\` but I don't know if that might cause a conflict.

**Comment by smiezah-msft** [comment]
Hi @iancward, I did some digging and it seems like that should not cause a conflict

**Comment by iancward** [comment]
@microsoft-github-policy-service agree company="The Walt Disney Company"

**Review by smiezah-msft — APPROVED** [review]
Thanks for the update. This approach aligns with the recommendation I was suggesting earlier.
